platforms/software/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/report/generic/MetadataRendererRegistry.java

```diff
@@ -19,75 +19,100 @@
-            return Object.class;
-        }
+public final class MetadataRendererRegistry {
+    private static final MetadataRenderer UNKNOWN_TYPE_RENDERER = new UnknownTypeRenderer();
 
-        @Override
-        public SimpleHtmlWriter render(Object metadata, SimpleHtmlWriter htmlWriter) throws IOException {
-            return (SimpleHtmlWriter) htmlWriter
-                .startElement("span").attribute("class", "unrenderable")
-                    .characters("[unrenderable type]")
-                .endElement();
-        }
-    };
-    private final LoadingCache<String, MetadataRenderer<?>> rendererLookupCache = CacheBuilder.newBuilder()
+    private final Set<MetadataRenderer> registeredRenderers = new HashSet<>();
+    private final LoadingCache<String, MetadataRenderer> rendererLookupCache = CacheBuilder.newBuilder()
         .maximumSize(100)
-        .build(new CacheLoader<String, MetadataRenderer<?>>() {
+        .build(new CacheLoader<String, MetadataRenderer>() {
             @Override
-            public MetadataRenderer<?> load(String metadataTypeName) {
+            public MetadataRenderer load(String metadataTypeName) {
                 return lookupRenderer(metadataTypeName);
             }
         });
 
     public MetadataRendererRegistry() {
-        registerRenderer(new StringMetadataRenderer());
-        registerRenderer(new NumberMetadataRenderer());
-        registerRenderer(new URIMetadataRenderer());
+        registerRenderer(new BasicRenderer());
+        registerRenderer(new ClickableLinkRenderer());
     }
 
-    public void registerRenderer(MetadataRenderer<?> metadataRenderer) {
-        renderers.add(metadataRenderer);
+    /**
+     * Registers a new {@link MetadataRenderer} instance with this registry.
+     * <p>
+     * The types returned by {@link MetadataRenderer#getMetadataTypes()} should not be assignable from any other
+     * types returned by the same method called on already registered renderers.
+     *
+     * @param metadataRenderer the renderer to register
+     */
+    public void registerRenderer(MetadataRenderer metadataRenderer) {
+        registeredRenderers.add(metadataRenderer);
     }
 
-    @SuppressWarnings("unchecked")
-    public <T> MetadataRenderer<T> getRenderer(String metadataTypeName) {
-        return (MetadataRenderer<T>) rendererLookupCache.getUnchecked(metadataTypeName);
+    /**
+     * Returns the {@link MetadataRenderer} instance that can render the given type of metadata.
+     *
+     * @param metadataTypeName the fully qualified name of the metadata type
+     * @return the renderer for the given type
+     */
+    public MetadataRenderer getRenderer(String metadataTypeName) {
+        return rendererLookupCache.getUnchecked(metadataTypeName);
     }
 
-    @SuppressWarnings("unchecked")
-    private <T> MetadataRenderer<T> lookupRenderer(String metadataTypeName) {
-        Class<T> type;
+    private MetadataRenderer lookupRenderer(String metadataTypeName) {
```

Comment:
Why keep this in the main class instead of inlining it to the cache? When reading code I find it cleaner if the loading part is defined where it is used, and it's not so long that it might be hard to read with extra indents.

Reply:
No good reason, I just didn't think about it after I got it working.  Will relocate.